src/workbase.rs

```diff
@@ -0,0 +1,27 @@
+use super::*;
+
+#[derive(Clone, Debug)]
+pub struct Workbase {
+    template: BlockTemplate,
+    merkle_branches: Vec<MerkleNode>,
+}
+
+impl Workbase {
+    pub fn new(template: BlockTemplate) -> Self {
+        let merkle_branches =
+            stratum::merkle_branches(template.transactions.iter().map(|tx| tx.txid).collect());
+
+        Self {
+            template,
+            merkle_branches,
+        }
+    }
```

Comment:
Lets wrap the new `Workbase` into `Arc` for sharing across Jobs, since multiple Jobs with different extranonces should share the same `Workbase`. 

This will make sure the type system enforces sharing by default, it also prevents accidental non-shared usage and clears our API contract a bit.


```suggestion
impl Workbase {
    pub fn new(template: BlockTemplate) -> Arc<Self> {
        let merkle_branches = stratum::merkle_branches(
            template.transactions.iter().map(|tx| tx.txid).collect()
        );
        Arc::new(Self { template, merkle_branches })
    }
}
```